v2/tools/generator/internal/astmodel/meta_type.go

```diff
@@ -161,3 +161,21 @@ func ExtractTypeName(aType Type) (TypeName, bool) {
 
 	return TypeName{}, false
 }
+
+// ExtractPrimitiveType extracts a PrimitiveType from the specified type if possible. This includes unwrapping
+// MetaType's like ValidatedType as well as checking the element type of types such as ArrayType and MapType.
+func ExtractPrimitiveType(aType Type) (*PrimitiveType, bool) {
+	if primitiveType, ok := AsPrimitiveType(aType); ok {
+		return primitiveType, ok
+	}
+
+	if arrayType, ok := AsArrayType(aType); ok {
+		return ExtractPrimitiveType(arrayType.Element())
+	}
+
+	if mapType, ok := AsMapType(aType); ok {
+		return ExtractPrimitiveType(mapType.ValueType())
```

Comment:
Looking only at the value type of the map might be surprising to future consumers, so I'm wondering whether this method belongs here.

Either move it closer to the (only) usage, so it's clear it's specialised, OR add a comment to the code indicating why only the map value needs to be checked.

Reply:
This is a fair point. The reason I put this here is because it's structurally very similar to `ExtractTypeName` - which also has the "only checks map values" aspect. But... I also wrote that function and put it here, and it also only has a single usage (in the embedded resource handling).

I've moved the new function closer to its usage and removed it from here. Possibly the same should happen for `ExtractTypeName`, but I haven't moved it for now. Let me know if you think it should be moved too - although in its case the only checking map values has more validity as TypeNames are never the generated maps key.